### PR TITLE
Add Windows event logs to flare

### PR DESCRIPTION
**Pull request by ghost**
### What does this PR do?

Add Windows event logs to flare

### Motivation

To make flare information more useful to support professionals

### Additional Notes

No

### Describe your test plan

Run "agent flare" to create a flare and observe event logs are in the ZIP file. Analyze the log file to ensure event logs are exported properly.

### CODE REVIEW

**Comment by mikezhu-dd** [comment]
Agent.exe doesn't have permission to export all event log channels. If we want absolutely all logs, we have to grant agent service more permission.

could not export event log Microsoft-Windows-AppXDeploymentServer/Restricted, error: Access is denied.
could not export event log Microsoft-Windows-ASN1/Operational, error: Access is denied.
could not export event log Microsoft-Windows-BranchCacheSMB/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-CAPI2/Operational, error: Access is denied.
could not export event log Microsoft-Windows-Crypto-BCRYPT/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-Crypto-CNG/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-Crypto-DSSEnh/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-Crypto-NCrypt/Operational, error: Access is denied.
could not export event log Microsoft-Windows-Crypto-RNG/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-Crypto-RSAEnh/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-GroupPolicy/Operational, error: Access is denied.
could not export event log Microsoft-Windows-Help/Operational, error: Access is denied.
could not export event log Microsoft-Windows-LiveId/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-LiveId/Operational, error: Access is denied.
could not export event log Microsoft-Windows-NetworkProvider/Operational, error: Access is denied.
could not export event log Microsoft-Windows-OfflineFiles/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-OfflineFiles/Debug, error: Access is denied.
could not export event log Microsoft-Windows-OfflineFiles/SyncLog, error: Access is denied.
could not export event log Microsoft-Windows-SMBClient/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-SmbClient/Audit, error: Access is denied.
could not export event log Microsoft-Windows-SmbClient/Connectivity, error: Access is denied.
could not export event log Microsoft-Windows-SMBClient/Operational, error: Access is denied.
could not export event log Microsoft-Windows-SmbClient/Security, error: Access is denied.
could not export event log Microsoft-Windows-SMBServer/Analytic, error: Access is denied.
could not export event log Microsoft-Windows-SMBServer/Audit, error: Access is denied.
could not export event log Microsoft-Windows-SMBServer/Connectivity, error: Access is denied.
could not export event log Microsoft-Windows-SMBServer/Operational, error: Access is denied.
could not export event log Microsoft-Windows-SMBServer/Performance, error: Access is denied.
could not export event log Microsoft-Windows-SMBServer/Security, error: Access is denied.
could not export event log Microsoft-Windows-StateRepository/Restricted, error: Access is denied.
could not export event log Microsoft-Windows-UAC-FileVirtualization/Operational, error: Access is denied.
could not export event log Microsoft-Windows-Windows Firewall With Advanced Security/ConnectionSecurity, error: Access is denied.
could not export event log Microsoft-Windows-Windows Firewall With Advanced Security/ConnectionSecurityVerbose, error: Access is denied.
could not export event log Network Isolation Operational, error: Access is denied.

File: pkg/flare/archive.go [hunk 1]
```diff
@@ -225,11 +225,17 @@ func createArchive(zipFilePath string, local bool, confSearchPaths SearchPaths,
 	if err != nil {
 		log.Errorf("Could not write typeperf data: %s", err)
 	}
+
 	err = zipCounterStrings(tempDir, hostname)
 	if err != nil {
 		log.Errorf("Could not write counter strings: %s", err)
 	}
 
+	err = zipEventLogs(tempDir, hostname)
```
Comment (hush-hush) [inline pkg/flare/archive.go:234]: Could we rename it to `zipWindowsEventLogs` ?

File: pkg/flare/archive_win.go [hunk 2]
```diff
@@ -82,6 +96,76 @@ func zipTypeperfData(tempDir, hostname string) error {
...
+func zipEventLogs(tempDir, hostname string) error {
+	// Enumerate all event channels registered.
+	eventLogChannels, err := windowsevent.EnumerateChannels()
+
+	if err != nil {
+		log.Warnf("could not list windows event log channels: %v", err)
+		return err
+	}
+
+	log.Infof("found event log channels: %v", eventLogChannels)
```
Comment (hush-hush) [inline pkg/flare/archive_win.go:109]: Should this be in `debug` ?

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/flare/archive.go (hunk 1 above)
Comment (mikezhu-dd) [reply pkg/flare/archive.go:234]: It follows the convention of zipCounterStrings. Won't change.

File: pkg/flare/archive_win.go (hunk 2 above)
Comment (mikezhu-dd) [reply pkg/flare/archive_win.go:109]: It need to appear in the final log file in flare. Info level is correct.

File: pkg/flare/archive.go (hunk 1 above)
Comment (hush-hush) [reply pkg/flare/archive.go:234]: I understand but `Logs` is a whole datadog product which makes it confusing I thinks

**Review by derekwbrown — CHANGES_REQUESTED** [review]
Should have a larger discussion, but IMO we should
- filter out the logs.  The PR as implemented dumps ALL of the event logs (of all types).  This can get quite large, AND might send us information that the customer doesn't want us to send. 

Suggestion would be only application/datadog, service control manager, and maybe WMI.

- Should actually do the conversion into text.  Including the logs as .evtx files means they _must_ be viewed on a windows machine, which will make it hard for the avg. support person to view.

**Comment by mikezhu-dd** [comment]
> Should have a larger discussion, but IMO we should
> 
> * filter out the logs.  The PR as implemented dumps ALL of the event logs (of all types).  This can get quite large, AND might send us information that the customer doesn't want us to send.
> 
> Suggestion would be only application/datadog, service control manager, and maybe WMI.
> 
> * Should actually do the conversion into text.  Including the logs as .evtx files means they _must_ be viewed on a windows machine, which will make it hard for the avg. support person to view.

Do we have a product manager or a PM who can define a feature? Is there a hard size limit for the flare file?

**Comment by olivielpeau** [comment]
A couple of notes, hope it helps:

* I believe the max size of the flare file is 50MB (Zendesk limit, I don't think we enforce our own)
* I don't know if the event logs we're adding may contain credentials, but we need to make sure any credential is scrubbed/removed from what we're adding to the flare file

**Comment by mikezhu-dd** [comment]
> A couple of notes, hope it helps:
> 
> * I believe the max size of the flare file is 50MB (Zendesk limit, I don't think we enforce our own)
> * I don't know if the event logs we're adding may contain credentials, but we need to make sure any credential is scrubbed/removed from what we're adding to the flare file

Thanks for the info. I'm going to change the PR to only include certain set of events. Event logs don't have credentials. By default, all Users, rather than just Administrators, can read event log. Exposing credentials in event log would be considered as security issue.

**Review by hush-hush — APPROVED** [review]

**Comment by derekwbrown** [comment]
> * Should actually do the conversion into text.  Including the logs as .evtx files means they _must_ be viewed on a windows machine, which will make it hard for the avg. support person to view.

Still needs to convert to text.  It's unlikely support folks will be able to read .evtx files.
The api exists to convert to text/xml.

**Review by albertvaka — APPROVED** [review]
The conversion to text would be useful but we shouldn't block the PR on it. We can do it later.

**Pull request merged**